### PR TITLE
add setting to open folders in a new window, fixes #371

### LINKED ISSUE #371: Mac: File -> Open should open in new window
On Mac, the standard behavior of the File -> Open menu is to open in a new window, but Visual Studio Code instead replaces the project in the current window with the new one. Achieving the standard Mac behavior requires first selecting File -> New Window before selecting File -> Open. Please instead make File -> Open follow the standard behavior.

### PR DESCRIPTION
**Pull request by elemongw**
I used the same logic used for the window.openFilesInNewWindow setting

### CODE REVIEW

**Comment by elemongw** [comment]
This doesn't work correctly yet and unfortunately I run out of time to investigate. When I was running it
`const windowConfig = this.configurationService.getConfiguration<IWindowSettings>('window');`
was always null. I'm probably doing something wrong. If no one steps up I will at this again next weekend.

File: src/vs/code/electron-main/menus.ts [hunk 1]
```diff
@@ -438,8 +438,7 @@ export class VSCodeMenu {
 	private createOpenRecentMenuItem(path: string): Electron.MenuItem {
 		return new MenuItem({
 			label: unMnemonicLabel(path), click: (menuItem, win, event) => {
-				const openInNewWindow = event && ((!platform.isMacintosh && event.ctrlKey) || (platform.isMacintosh && event.metaKey));
-				const success = !!this.windowsService.open({ cli: this.environmentService.args, pathsToOpen: [path], forceNewWindow: openInNewWindow });
+				const success = !!this.windowsService.open({ cli: this.environmentService.args, pathsToOpen: [path] });
```
Comment (bpasero) [inline src/vs/code/electron-main/menus.ts]: @elemongw can we keep the semantic of forcing a new window when the user presses the Ctrl/Meta key? I see no reason to remove it.

File: src/vs/code/electron-main/windows.ts [hunk 2]
```diff
@@ -709,7 +709,18 @@ export class WindowsManager implements IWindowsService {
 		}
 
 		// Handle folders to open
-		let openInNewWindow = openConfig.preferNewWindow || openConfig.forceNewWindow;
+		let openInNewWindow: boolean;
```
Comment (bpasero) [inline src/vs/code/electron-main/windows.ts]: @elemongw sorry, but this does not work for me when I try it. The folder opens in the same window no matter how I configure the setting. I believe you need to turn the logic around: set `openInNewWindow` to true when the setting is enabled!

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
@elemongw added some feedback, I think your code is never really executed because `openConfig.preferNewWindow` is never true for folders. Please revisit your logic there.

Nevertheless, calling `this.configurationService.getConfiguration<IWindowSettings>('window')` returns something for me IF the setting is defined in `settings.json`. Otherwise you will get undefined because there are no settings. In that case you just have to pick a default (in this case the default is also a falsify value because by default folders open in the same window. So you will have to check for `window` being undefined and `window.openFoldersInNewWindow` be `undefined` or `false`.

**Comment by bpasero** [comment]
@elemongw ping

**Comment by elemongw** [comment]
Sorry I've been very busy, I'll try to work on this over the weekend.

**Comment by bpasero** [comment]
No worries, was just wondering if you got my notifications 👍

**Comment by elemongw** [comment]
I updated the patch. One question, is the intended behaviour for this.configurationService.getConfiguration<IWindowSettings>('window') to only contain options explicitly set by the user in the configuration file or should it also contain the default values?

**Comment by elemongw** [comment]
@bpasero ping

**Comment by bpasero** [comment]
Thanks for the reminder, I am planning to do a full review of how windows open and when across all OS but I did not have time to do it yet. When I do I will also revisit this PR 👍

**Comment by bpasero** [comment]
This landed in a slightly different form, please see https://github.com/Microsoft/vscode/issues/371

Thanks for this PR though, it made me rethink the whole way how files and folders open in VS Code 👍

**Pull request closed without merging**